src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+                schedule,
+                run_command(config.clone(), shutdown.clone(), out.clone()),
+            );
+
+            let timeout_result = timeout.await;
+
+            match timeout_result {
+                Ok(output) => {
+                    if let Err(command_error) = output {
+                        emit!(ExecFailed {
+                            command: config.command.as_str(),
+                            error: command_error,
+                        });
+                    }
+                }
+                Err(_) => {
+                    emit!(ExecTimeout {
+                        command: config.command.as_str(),
+                        elapsed_seconds: now.elapsed().as_secs(),
+                    });
+                }
+            }
+        }
+
+        info!("Finished scheduled exec runs.");
+        Ok(())
+    }))
+}
+
+pub fn run_streaming(
+    config: ExecConfig,
+    respawn_on_exit: bool,
+    respawn_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        if respawn_on_exit {
+            let duration = time::Duration::from_secs(respawn_interval_secs);
+
+            // Continue to loop while shutdown is pending
+            while futures::poll!(shutdown.clone()).is_pending() {
+                let output = run_command(config.clone(), shutdown.clone(), out.clone()).await;
+
+                if let Err(command_error) = output {
+                    emit!(ExecFailed {
+                        command: config.command.as_str(),
+                        error: command_error,
+                    });
+                }
+
+                if futures::poll!(shutdown.clone()).is_ready() {
+                    break;
+                } else {
+                    warn!("Streaming processed ended before shutdown.");
+
+                    // Using time interval so it can utilize the shutdown signal
+                    let mut interval = time::interval(duration).take_until(shutdown.clone());
+
+                    // Call next twice since the first is immediate
+                    interval.next().await;
+                    interval.next().await;
+                }
+            }
+        } else {
+            let output = run_command(config.clone(), shutdown, out).await;
+
+            if let Err(command_error) = output {
+                emit!(ExecFailed {
+                    command: config.command.as_str(),
+                    error: command_error,
+                });
+            }
+        }
+
+        Ok(())
+    }))
+}
+
+async fn run_command(
+    config: ExecConfig,
+    shutdown: ShutdownSignal,
+    mut out: Pipeline,
+) -> Result<Option<ExitStatus>, Error> {
+    info!("Starting command run.");
+    let mut command = build_command(&config);
+
+    // Mark the start time just before spawning the process as
+    // this seems to be the best approximation of exec duration
+    let now = Instant::now();
```

Comment:
Nit: I suggest naming this `start` to be clear to its purpose.

Reply:
Fixed